python/scrutiny/server/device/request_dispatcher.py

```diff
@@ -1,59 +1,130 @@
 #    request_dispatcher.py
-#        Use a PriorityQueue to dispatch Request to the device. Associate each request wil
+#        Use a PriorityQueue to dispatch Request to the device. Associate each request with
 #        its callback
 #
 #   - License : MIT - See LICENSE file.
 #   - Project : Scrutiny Debugger (github.com/scrutinydebugger/scrutiny)
 #
 #   Copyright (c) 2021-2022 scrutinydebugger
 
-import queue
+import bisect
+from scrutiny.server.protocol import Response
+from scrutiny.server.server_tools import Throttler
+from time import time
+import math
+import logging
 
 
-class RequestDispatcher:
+class RequestQueue:
+    """
+    Non-thread-safe Queue with priority.
+    Replace queue.PriorityQueue simply because I don't like that they compare data and don't want to introduce workarounds or 
+    dataclass from Python 3.7 just for not comparing the data when selecting priority.
+    We will have all the flexibility we need with our own minimalist custom class
+    """
+
+    def __init__(self, size=None):
+        self.clear()
+        self.size = size
+
+    def clear(self):
+        self.data = []
+        self.priorities = []
 
-    class RequestRecord:
-        __slots__ = ('request', 'success_callback', 'failure_callback', 'success_params', 'failure_params', 'completed')
+    def push(self, item, priority=0):
+        if self.size is not None and len(self.data) >= self.size:
+            raise Exception('Request queue full')
 
-        def __init__(self):
-            self.completed = False
+        index = bisect.bisect_left(self.priorities, priority)
+        self.data.insert(index, item)
+        self.priorities.insert(index, priority)
 
-        def complete(self, success=False, response=None, response_data=None):
-            self.completed = True  # Set to true at beginning so that it is still true if an exception raise in the callback
-            if success:
-                if response is None or response_data is None:
-                    raise ValueError('Missing response')
-                self.success_callback(self.request, response.code, response_data, self.success_params)
-            else:
-                self.failure_callback(self.request, self.failure_params)
+    def pop(self):
+        if len(self.data) > 0:
+            item = self.data[-1]
+            del self.priorities[-1]
+            del self.data[-1]
+            return item
 
-        def is_completed(self):
-            return self.completed
+    def peek(self):
+        if len(self.data) > 0:
+            return self.data[-1]
 
-        # Workaround for PriorityQueue that compare the data when priority is equal. bpo-31145
```

Comment:
See PriorityQueue documentation on how to avoid that.
Indeed overriding internal python methods is a red flags.

Reply:
This has been removed. It's an old commit